core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +241,120 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
+   * network issues, the HeartbeatReceiver will put the executor into `executorExpiryCandidates`
+   * rather than expiring it immediately.
+   */
     logTrace("Checking for hosts with no recent heartbeats in HeartbeatReceiver.")
+    logWarning(s"Keep `expiryCandidatesTimeout` larger than `HEARTBEAT_MILLIS` in" +
+      s"deploy/worker/Worker.scala to know whether master lost any heartbeat from the" +
+      s"worker or not.")
     val now = clock.getTimeMillis()
-    for ((executorId, lastSeenMs) <- executorLastSeen) {
-      if (now - lastSeenMs > executorTimeoutMs) {
-        logWarning(s"Removing executor $executorId with no recent heartbeats: " +
-          s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
-        // Asynchronously kill the executor to avoid blocking the current thread
-        killExecutorThread.submit(new Runnable {
-          override def run(): Unit = Utils.tryLogNonFatalError {
-            // Note: we want to get an executor back after expiring this one,
-            // so do not simply call `sc.killExecutor` here (SPARK-8119)
-            sc.killAndReplaceExecutor(executorId)
-            // SPARK-27348: in case of the executors which are not gracefully shut down,
-            // we should remove lost executors from CoarseGrainedSchedulerBackend manually
-            // here to guarantee two things:
-            // 1) explicitly remove executor information from CoarseGrainedSchedulerBackend for
-            //    a lost executor instead of waiting for disconnect message
-            // 2) call scheduler.executorLost() underlying to fail any tasks assigned to
-            //    those executors to avoid app hang
-            sc.schedulerBackend match {
-              case backend: CoarseGrainedSchedulerBackend =>
-                backend.driverEndpoint.send(RemoveExecutor(executorId,
-                  ExecutorProcessLost(
-                    s"Executor heartbeat timed out after ${now - lastSeenMs} ms")))
-
-              // LocalSchedulerBackend is used locally and only has one single executor
-              case _: LocalSchedulerBackend =>
-
-              case other => throw new UnsupportedOperationException(
-                s"Unknown scheduler backend: ${other.getClass}")
-            }
+    if (!checkWorkerLastHeartbeat || !isStandalone()) {
+      for ((executorId, lastSeenMs) <- executorLastSeen) {
+        if (now - lastSeenMs > executorTimeoutMs) {
+          killExecutor(executorId, now - lastSeenMs)
+          executorLastSeen.remove(executorId)
+        }
+      }
+    } else {
+      for ((executorId, workerLastHeartbeat) <- executorExpiryCandidates) {
+        if (now - workerLastHeartbeat > expiryCandidatesTimeout) {
+          killExecutor(executorId, now - workerLastHeartbeat)
+          executorExpiryCandidates.remove(executorId)
+          executorLastSeen.remove(executorId)
+        }
+      }
```

Comment:
Could we move this block to the end of this `else` branch? And then we don't need to kill executors while handling `workerLastHeartbeats` below but do all kills in the same block. E.g.,

```scala
val buf = new ArrayBuffer[String]()
for ((executorId, lastSeenMs) <- executorLastSeen) {
  if (now - lastSeenMs > executorTimeoutMs) {
      buf += executorId
  }
}
sc.schedulerBackend match {
  case backend: StandaloneSchedulerBackend =>
    backend.client.workerLastHeartbeat(sc.applicationId, buf) match {
      case Some(workerLastHeartbeats) =>
        for ((executorId, workerLastHeartbeat) <- buf zip workerLastHeartbeats) {
           executorExpiryCandidates(executorId) = workerLastHeartbeat
           executorLastSeen.remove(executorId)
        }
       case None =>
         for (executorId <- buf) {
           val lastSeenMs = executorLastSeen.get(executorId).get
           executorExpiryCandidates(executoId) =Long.Max
         }

for ((executorId, workerLastHeartbeat) <- executorExpiryCandidates) {
  if (now - workerLastHeartbeat > expiryCandidatesTimeout) {
   // One problem here might be that  we can not use `workerLastHeartbeat` for the None case as it was 
   // lastSeenMs previously. 
    killExecutor(executorId, now - workerLastHeartbeat)
    executorExpiryCandidates.remove(executorId)
    executorLastSeen.remove(executorId)
  }
}
```

Reply:
Updated https://github.com/apache/spark/pull/37411/commits/e9fe8d1db2bc5307a437e903d0991c3f4c7d1949

There is a slight difference when `backend.client.workerLastHeartbeat` matches the case `None`. I set `executorExpiryCandidates(executorId)` to `lastSeenMs` instead of `Long.Max`.